### DISCUSSION TITLE
Possible Compiler Bug RE constraints

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5054] [Q&A] [answered]
The code below seems to me that it should compile but it does not.
Here's what seems wonky:
```MyBase``` specifies a constraint on ```Parse```
```MyClass``` implements the ```Parse``` function
You get a compiler error saying you can't specify constraints (other than class or struct) if you uncomment ```where T : DisplayRecord```
You get a compiler errors if no constraints are specified on ```MyClass```'s ```Parse```.
The code only compiles if you specify ```where T : class``` on ```MyClass``` but this seems as though it shouldn't be necessary because ```T : DisplayRecord``` and ```DisplayRecord : record``` and ```record : class```.

Thoughts?

```
#nullable enable

using System;

namespace ConsoleApp1
{
    
    public class MyClass : MyBase
    {
        public override T? Parse<T>(string Text)
            //where T : DisplayRecord
            //where T: class

        {
            return base.Parse<T>(Text);
        }
    }

    public abstract class MyBase
    {
        public virtual T? Parse<T>(string Text) 
            where T : DisplayRecord
        {
            throw new NotImplementedException();
        }
    }

    public record DisplayRecord
    {

    }

}

```

**Comment by HaloFour** [comment, marked as answer]
Are you suggesting that this is a bug in the specification, or a bug in the compiler?  If it's the latter it should probably be reported to the Roslyn repo.

**Reply by TonyValenti** [reply]
@HaloFour - I'm not sure but it seems to me the above should compile without needing to specify ````T : class```.
I'll reopen in Roslyn.

**Reply by jnm2** [reply]
I believe you're missing `where T : default`. Overrides require it when `T?` is used in the signature of the base method and `T` is not constrained to a class or struct. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters#default-constraint

**Reply by svick** [reply]
@jnm2 [This code doesn't compile with `where T : default`.](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4GIB2BXANlghgIywgAII0CiBYAKBoAEAmARhpvIFsIBnAB1wGMSAYQD2aLiKIBBHjxbUA3jWIriy1XQDMxRsQCyCIXi5diIfQgBCuLhHUql1Vc53aRkKLAAmJACoB+YgAFXChbAB5fAD4ACjomAAZiXwgAD2AASnsXZwB6XIB3AAtoPzNiABEYXjwEACUIfhEoL2yclWLS5PMfOFxsYFYndsd2lzoAdmJ8GwgAOhCwiEjYlPSMgG421QBfbL3aYZUtYgIuYCgBYB0GC2tbbNH2k7AYKGAMXCxkwMWI6LiiWSaUyaiOY2InSgZXMVRquHqjWarXBzieEOIwCKUBEBWIaAgeIAciJgABJdg8IicNDACBeACiqUEPGAMDEMU22xUB2cB2yJ2hTRalWqVIRDWFKLRQz5QwOQA==) It requires `where T : class`, because the base method is already constrained to a class.
